[inductor] overlap passes, bucket first exposed collectives

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #169942


Optimization for scarce compute situation.

Exposed collectives are on critical path, bucketing them will directly reduce critical path via amortizing communication cost etc.

For torchtitan llama3 FSDP+TP shows 16 -> 23 mfu.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

IvanKobzarev committed:
[inductor] overlap passes, bucket first exposed collectives

[ghstack-poisoned]

eellison commented on torch/_inductor/fx_passes/overlap_preserving_bucketer.py:
```diff
@@ -356,8 +367,14 @@ def _find_buckets(
+        # Exposed collectives benefit most from bucketing since their latency is on the
+        # critical path. Prioritizing them also preserves hiding relationships for
+        # already-hidden collectives, which have less to gain from bucketing.
+        sorted_collectives = sorted(
```
Hmm, could we do something where we sort first by exposure, but then try to bucket locally ? it feels a bit unintuitive to try to bucket from the start of the graph till the end.

See also:

```
            # Greedy optimization: stop after consecutive failures
            consecutive_failures = 0
            max_consecutive_failures = 20

            # Check candidates in sorted order, break when beyond max distance
            for candidate in sorted_collectives[i + 1 : i + 1 + self.max_coll_distance]:
```

IvanKobzarev committed:
Update on "[inductor] overlap passes, bucket first exposed collectives"



Optimization for scarce compute situation.

Exposed collectives are on critical path, bucketing them will directly reduce critical path via amortizing communication cost etc.

For torchtitan llama3 FSDP+TP shows 16 -> 23 mfu.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

eellison commented on torch/_inductor/fx_passes/overlap_preserving_bucketer.py:
```diff
@@ -396,6 +404,7 @@ def _find_buckets(
             # Greedy optimization: stop after consecutive failures
             consecutive_failures = 0
             max_consecutive_failures = 20
+            start_node_idx = self.node_idx[start_node]
```
can we iterate in local vicinity first ? e.g 1, then -1, then 2, then -2. 

The reason we dont do that previously is we start at beginning, so we dont need to try to bucket nodes that have already been attempted

IvanKobzarev replied:
From offline discussion: decided to keep bucketing in more exposed first order with node distance limitation to get more signal about performance on different models.

IvanKobzarev committed:
Update on "[inductor] overlap passes, bucket first exposed collectives"



Optimization for scarce compute situation.

Exposed collectives are on critical path, bucketing them will directly reduce critical path via amortizing communication cost etc.

For torchtitan llama3 FSDP+TP shows 16 -> 23 mfu.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[inductor] overlap passes, bucket first exposed collectives"



Optimization for scarce compute situation.

Exposed collectives are on critical path, bucketing them will directly reduce critical path via amortizing communication cost etc.

For torchtitan llama3 FSDP+TP shows 16 -> 23 mfu.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[inductor] overlap passes, bucket first exposed collectives"



Optimization for scarce compute situation.

Exposed collectives are on critical path, bucketing them will directly reduce critical path via amortizing communication cost etc.

For torchtitan llama3 FSDP+TP shows 16 -> 23 mfu.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[inductor] overlap passes, bucket first exposed collectives"



Optimization for scarce compute situation.

Exposed collectives are on critical path, bucketing them will directly reduce critical path via amortizing communication cost etc.

For torchtitan llama3 FSDP+TP shows 16 -> 23 mfu.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]
